contracts/SDUtilityPool.sol

```diff
@@ -0,0 +1,771 @@
+        _delegate(sdAmount);
+    }
+
+    /**
+     * @notice auxiliary method to put a withdrawal request, takes in cToken in input
+     * @param _cTokenAmount amount of cToken
+     * @return _requestId generated request ID for withdrawal
+     */
+    function requestWithdraw(uint256 _cTokenAmount) external override whenNotPaused returns (uint256 _requestId) {
+        if (_cTokenAmount > delegatorCTokenBalance[msg.sender]) {
+            revert InvalidAmountOfWithdraw();
+        }
+        accrueFee();
+        uint256 exchangeRate = _exchangeRateStoredInternal();
+        uint256 sdRequested = (exchangeRate * _cTokenAmount) / DECIMAL;
+        _requestId = _requestWithdraw(sdRequested, _cTokenAmount);
+    }
+
+    /**
+     * @notice auxiliary method to put a withdrawal request, takes SD amount in input
+     * @param _sdAmount amount of SD to withdraw
+     * @return _requestId generated request ID for withdrawal
+     */
+    function requestWithdrawWithSDAmount(uint256 _sdAmount)
+        external
+        override
+        whenNotPaused
+        returns (uint256 _requestId)
+    {
+        accrueFee();
+        uint256 exchangeRate = _exchangeRateStoredInternal();
+        uint256 cTokenToBurn = (_sdAmount * DECIMAL) / exchangeRate;
+        if (cTokenToBurn > delegatorCTokenBalance[msg.sender]) {
+            revert InvalidAmountOfWithdraw();
+        }
+        _requestId = _requestWithdraw(_sdAmount, cTokenToBurn);
+    }
+
+    /**
+     * @notice finalize delegator's withdraw requests
+     */
+    function finalizeDelegatorWithdrawalRequest() external override whenNotPaused {
+        accrueFee();
+        uint256 exchangeRate = _exchangeRateStoredInternal();
+        uint256 maxRequestIdToFinalize = Math.min(nextRequestId, nextRequestIdToFinalize + finalizationBatchLimit) - 1;
+        uint256 requestId;
+        uint256 sdToReserveToFinalizeRequest;
+        for (requestId = nextRequestIdToFinalize; requestId <= maxRequestIdToFinalize; ) {
+            DelegatorWithdrawInfo memory delegatorWithdrawInfo = delegatorWithdrawRequests[requestId];
+            uint256 requiredSD = delegatorWithdrawInfo.sdExpected;
+            uint256 amountOfcToken = delegatorWithdrawInfo.amountOfCToken;
+            uint256 minSDRequiredToFinalizeRequest = Math.min(requiredSD, (amountOfcToken * exchangeRate) / DECIMAL);
+            if (
+                (sdToReserveToFinalizeRequest + minSDRequiredToFinalizeRequest > getPoolAvailableSDBalance()) ||
+                (delegatorWithdrawInfo.requestBlock + minBlockDelayToFinalizeRequest > block.number)
+            ) {
+                break;
+            }
+            delegatorWithdrawRequests[requestId].sdFinalized = minSDRequiredToFinalizeRequest;
+            sdRequestedForWithdraw -= requiredSD;
+            sdToReserveToFinalizeRequest += minSDRequiredToFinalizeRequest;
+            delegatorCTokenBalance[delegatorWithdrawInfo.owner] -= amountOfcToken;
```

Comment:
Shouldn't this be done during request withdrawal? Otherwise, user will be able to create multiple requests reusing the same cTokens. This might block the entire finalization flow for all users.

Reply:
If doing earlier poses a problem with exchange rate, we should at least separate this variable into a new mapping for (address => withdrawStageCTokenCount)

Reply:
good catch. Yeah, it will impact the exchange rate. I will push the fix with the new mapping of withdrawCTokenCount